builtin/tfc/config_sourcer.go

```diff
@@ -0,0 +1,478 @@
+				}
+
+				continue
+			}
+
+			id = data.Data.Id
+
+			if cs.workspaceIds == nil {
+				cs.workspaceIds = map[string]string{}
+			}
+
+			L.Info("mapped Terraform Cloud organization to workspace-id",
+				"organization", tfcReq.Organization,
+				"workspace", tfcReq.Workspace,
+				"workspace-id", id)
+
+			cs.workspaceIds[key] = id
+		}
+
+		// Get this secret or read it if we haven't already.
+		cachedSecretVal, ok := cs.secretCache[key]
+		if !ok {
+			L.Trace("querying TFC workspace")
+
+			url := fmt.Sprintf("%s/api/v2/workspaces/%s/current-state-version?include=outputs",
+				baseURL, id)
+
+			req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+			if err != nil {
+				L.Error("error in creating request for state version", "error", err)
+
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, err.Error()).Proto(),
+				}
+
+				continue
+			}
+
+			req.Header.Set("Authorization", "Bearer "+cs.config.Token)
+			req.Header.Set("Content-Type", "application/vnd.api+json")
+
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				L.Error("error in sending request for state version", "error", err)
+
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, err.Error()).Proto(),
+				}
+
+				continue
+			}
+
+			defer resp.Body.Close()
+
+			var data stateLookup
+
+			err = json.NewDecoder(resp.Body).Decode(&data)
+			if err != nil {
+				L.Error("error in decoding outputs", "error", err)
+
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, err.Error()).Proto(),
+				}
+
+				continue
+			}
+
+			outputs := map[string]string{}
+
+			for _, output := range data.Included {
+				if output.Type != "state-version-outputs" {
+					L.Info("ignored unexpected type in included values", "type", output.Type)
+					continue
+				}
+
+				outputs[output.Attributes.Name] = output.Attributes.Value
+			}
+
+			L.Info("refreshed outputs from Terraform Cloud", "workspace", id, "vars", len(outputs))
+
+			cachedSecretVal = &cachedSecret{Outputs: outputs}
+			cs.secretCache[id] = cachedSecretVal
+		}
+
+		value := cachedSecretVal.Outputs[tfcReq.Output]
+
+		result.Result = &pb.ConfigSource_Value_Value{
+			Value: value,
+		}
+	}
+
+	return results, nil
+}
+
+func (cs *ConfigSourcer) stop() error {
+	cs.cacheMu.Lock()
+	defer cs.cacheMu.Unlock()
+
+	if cs.client != nil {
+		cs.client.CloseIdleConnections()
+		cs.client = nil
+	}
+
+	// Reset our results tracking to empty. This will force the next call
+	// to rebuild all our secret values.
+	var zeroTime time.Time
+	cs.lastRead = zeroTime
+	cs.secretCache = nil
+
+	return nil
+}
+
+func (cs *ConfigSourcer) Documentation() (*docs.Documentation, error) {
+	doc, err := docs.New(
+		docs.FromConfig(&sourceConfig{}),
+		docs.RequestFromStruct(&reqConfig{}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	doc.Description("Read configuration values from Terraform Cloud.")
```

Comment:
This doesn't feel exactly right, or perhaps could be worded differently for TFC. "Read Terraform output values from Terraform Cloud." or something I dunno.